(feat) Edit Patient

MatthewDorner opened the pull request:
Fixes #1765 

- Implement Edit Patient functionality
- Refactor New Patient and View Patient so that New, Edit and View Patient functions all use the same `GeneralInformation` form.
- Changed behavior of `isEditable` prop in input components to make it false by default, as this seems to make more sense for boolean
- Moved `createPatient` into patient slice instead of patients

Right now the tests should all pass but I still need to add the tests for Edit Patient and improve some other tests. I think I also need to add error handling back into the new form since I lost it when I deleted `NewPatientForm`.

MatthewDorner commented on src/clients/db/Repository.ts:
```diff
@@ -57,16 +57,15 @@ export default class Repository<T extends AbstractDBModel> {
         _rev: rev,
-        ...dataToSave,
+        ...restOfDoc,
+        ...entity,
```
don't remember changing this...

MatthewDorner replied:
yeah accidentally overwrote your fix  for the duplicate patients, should be good now

MatthewDorner commented on src/patients/related-persons/RelatedPersonTab.tsx:
```diff
@@ -49,20 +51,16 @@ const RelatedPersonTab = (props: Props) => {
+    }
+
+    patientToUpdate.relatedPersons.push(relatedPerson)
+    dispatch(updatePatient(patientToUpdate, history))
```
don't remember changing this either, must've messed it up in git

jackcmeyer commented on src/__tests__/patients/GeneralInformation.test.tsx:
```diff
@@ -21,95 +23,76 @@ describe('General Information', () => {
-    const wrapper = setup()
-
-    const dateOfBirthInput = wrapper.findWhere((w) => w.prop('name') === 'address')
+    const dateOfBirthInput = wrapper.findWhere((w: any) => w.prop('name') === 'address')
```
I think this variable name should be `addressTextField`. This is probably my bad, copy paste :(

MatthewDorner replied:
done

jackcmeyer commented on src/__tests__/patients/GeneralInformation.test.tsx:
```diff
@@ -21,95 +23,76 @@ describe('General Information', () => {
-    const wrapper = setup()
-
-    const dateOfBirthInput = wrapper.findWhere((w) => w.prop('name') === 'occupation')
+    const dateOfBirthInput = wrapper.findWhere((w: any) => w.prop('name') === 'occupation')
```
this should probably be `occupationInput`

MatthewDorner replied:
done

jackcmeyer commented on src/__tests__/patients/GeneralInformation.test.tsx:
```diff
@@ -21,95 +23,76 @@ describe('General Information', () => {
-    const wrapper = setup()
-
-    const dateOfBirthInput = wrapper.findWhere((w) => w.prop('name') === 'preferredLanguage')
+    const dateOfBirthInput = wrapper.findWhere((w: any) => w.prop('name') === 'preferredLanguage')
```
`preferredLanguageInput`

MatthewDorner replied:
done

jackcmeyer commented on src/__tests__/patients/GeneralInformation.test.tsx:
```diff
@@ -21,95 +23,76 @@ describe('General Information', () => {
-    const wrapper = setup()
-
-    const dateOfBirthInput = wrapper.findWhere((w) => w.prop('name') === 'phoneNumber')
+    const dateOfBirthInput = wrapper.findWhere((w: any) => w.prop('name') === 'phoneNumber')
```
`phoneNumberInput`

MatthewDorner replied:
done

jackcmeyer commented on src/HospitalRun.tsx:
```diff
@@ -42,6 +43,15 @@ const HospitalRun = () => {
+                  exact
+                  path="/patients/edit/:id"
+                  component={EditPatient}
+                />
```
I think that we need some tests in `HospitalRun.test.tsx` surrounding this new route, especially for permissions.

MatthewDorner replied:
Added these

MatthewDorner commented on src/__tests__/patients/patient-slice.test.ts:
```diff
@@ -130,6 +237,7 @@ describe('patients slice', () => {
     })
   })
 
+  // should check for the Toast
```
yeah. I added a toast for updating a Patient and made it redirect history to `/patients/${newPatient.id}`, since saving on Edit Patient ought to redirect to that. as a side effect of that, adding a new Related Person now moves you off the Related Persons tab and back onto the General Information tab, but I'm not sure if should do anything else about that now.

MatthewDorner replied:
added this test

MatthewDorner commented on src/patients/view/ViewPatient.tsx:
```diff
@@ -3,12 +3,13 @@ import { useDispatch, useSelector } from 'react-redux'
 import { useParams, withRouter, Route, useHistory, useLocation } from 'react-router-dom'
 import { Panel, Spinner, TabsHeader, Tab } from '@hospitalrun/components'
 import { useTranslation } from 'react-i18next'
+
```
just thought it helped to separate the types of imports

fox1t replied:
I really like separating type of imports:`external deps -> 1 empty line -> internal deps`
:)

MatthewDorner replied:
I think there's an eslint setting for this, that's why I instinctively did it.

MatthewDorner commented on src/patients/GeneralInformation.tsx:
```diff
@@ -273,27 +230,28 @@ const NewPatientForm = (props: Props) => {
-        )}
-      </form>
+      </Panel>
+      {isEditable ? (
```
Github thinks this file is NewPatientForm.tsx (which didn't have panels) renamed to GeneralInformation.tsx, when in fact I deleted NewPatientForm.tsx, so that's why it thinks I put panels in.

MatthewDorner commented on src/patients/edit/EditPatient.tsx:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  // see comment in ViewPatient
+  if (!reduxPatient) {
```
`isLoading` makes sense to me, but what does the check for `!reduxPatient.id` add? or should it be `!id`, like if there's no parameter? but that seems like it would be handled separately with an error or something

MatthewDorner replied:
so far I've just used `isLoading` which seems to work. Now, you actually see the loading icon when you change patients, etc.

MatthewDorner commented on src/patients/edit/EditPatient.tsx:
```diff
@@ -0,0 +1,86 @@
+  const dispatch = useDispatch()
+
+  const [patient, setPatient] = useState({} as Patient)
+  const { patient: reduxPatient } = useSelector((state: RootState) => state.patient)
```
done

MatthewDorner commented on src/patients/view/ViewPatient.tsx:
```diff
@@ -34,6 +37,8 @@ const ViewPatient = () => {
 
+  // this check doesn't work as an empty object isn't falsy. would it be more correct to display
+  // the spinner when the Redux patient state isLoading is true?
   if (!patient) {
```
in this one, test fails unless you use `isLoading || !patient` since it uses patient.id to render the tabs

MatthewDorner commented:
- Made the requested UI changes.
- Brought some of the tests from NewPatientForm.test.tsx back into GeneralInformation.test.tsx though probably still need more.
- Fixed other tests.
- Fixed the logic to display spinner in EditPatient and ViewPatient
- Added error handling and test for error handling back in.

Still need to add tests for the new route in HospitalRun component, and the actual tests for EditPatient.

jackcmeyer commented on src/patients/GeneralInformation.tsx:
```diff
@@ -273,27 +246,24 @@ const NewPatientForm = (props: Props) => {
-              <Button color="danger" onClick={() => onCancel()}>
-                {t('actions.cancel')}
-              </Button>
-            </div>
```
I think we want to keep these buttons

MatthewDorner replied:
that's where the buttons were in NewPatientForm, the entire file is deleted but Github thinks I renamed it to GeneralInformation and deleted GeneralInformation. The buttons are still there, they're just slightly further down. GeneralInformation has a different layout than NewPatientForm because it has the panels.

MatthewDorner replied:
ok, I restored the previous style of buttons

MatthewDorner commented:
Added tests for EditPatient component and the `patients/edit` route, still having some test failures and not exactly sure why. Also couldn't figure out how to mock `useParams` (see the commented out test in EditPatient.test.tsx)

MatthewDorner commented:
Oh, forgot the icon, we'll need to add a pencil icon in components.

tehKapa commented:
@MatthewDorner this afternoon we fixed an annoying bug with husky/lint. Could you Resolve conflicts and errors? thank you!

MatthewDorner commented on src/patients/GeneralInformation.tsx:
```diff
@@ -273,27 +246,27 @@ const NewPatientForm = (props: Props) => {
+                (event: React.ChangeEvent<HTMLTextAreaElement>) =>
+                  onFieldChange && onFieldChange('address', event.currentTarget.value)
+                // eslint-disable-next-line react/jsx-curly-newline
+              }
```
Can't get eslint to accept this any way I can think to format it, or via the eslint automatic fix. There must be two conflicting rules.

MatthewDorner commented:
- Fixed merge conflicts
- Moved buttons out of GeneralInformation form and into the parent forms (NewPatient, EditPatient, ViewPatient)
- Fixed some tests

MatthewDorner commented:
Tests passing now. Looks like I should add some more tests, although I was thinking of going through the tests more thoroughly after this gets merged anyway.

fox1t commented:
@MatthewDorner Great PR!

fox1t commented:
LGTM
